### ISSUE TITLE
Dimensionality of metrics

### ISSUE THREAD

**Issue by sachinruk** [keras-team/keras#7811]
I'm in the process of adding a new metric (and doing a pull request). However, I noticed that the dimensionality of how the metric operates isn't too obvious. So given a (6,7) matrix you end up with a vector of length 6 (see code below). As far as I understand the input matrix is of the shape, `(N_batch_size, num_features)`. So why are we evaluating the metric across the last dimension and not the first?

Also when I do `model.fit(x,y)` we always get one number for the metric. Even though according to the above I should be getting a `(N,)` shape vector. So the second question is, how do we get the final metric? Averaging it maybe?

From `metrics_test.py`:
```
def test_metrics():
    y_a = K.variable(np.random.random((6, 7)))
    y_b = K.variable(np.random.random((6, 7)))
    for metric in all_metrics:
        output = metric(y_a, y_b)
        print(metric.__name__)
        assert K.eval(output).shape == (6,)
```
